File: creational_design_patterns/singleton/ThreadSafeSingleton.py
import threading
from time import sleep

class ThreadSafeSingleton:
    _instance = None
    # lock object to synchronize threads during the first initialization
    _lock = threading.Lock()

    def __init__(self, data):
        # prevent direct instantiation from outside the class
        if ThreadSafeSingleton._instance is not None:
            raise Exception("This class is a singleton.")
        self.data = data

    @classmethod
    def get_instance(cls, data):
        # avoid acquiring the lock if the instance is already initialized
        if cls._instance is None:
            with cls._lock:  # synchronize threads
                # ensure only one thread creates the instance
                if cls._instance is None:
                    cls._instance = cls(data)
        return cls._instance

    def get_data(self):
        return self.data

def test_thread_safe_singleton():
    def create_singleton_instance(data):
        try:
            singleton = ThreadSafeSingleton.get_instance(data)
            print(f"Instance created with data: {singleton.get_data()}")
        except Exception as e:
            print(e)

    # Create two threads attempting to initialize the singleton with different data
    thread1 = threading.Thread(target=create_singleton_instance, args=("Thread 1 Data",))
    thread2 = threading.Thread(target=create_singleton_instance, args=("Thread 2 Data",))

    # Start the threads
    thread1.start()
    thread2.start()

    # Wait for threads to complete
    thread1.join()
    thread2.join()

    # Verify that the singleton instance is consistent
    singleton = ThreadSafeSingleton.get_instance("Main Data")
    print(f"Main thread singleton data: {singleton.get_data()}")

# Run the test
if __name__ == "__main__":
    test_thread_safe_singleton()

File: creational_design_patterns/singleton/ThreadSafeSingleton.kt
package singleton

class ThreadSafeSingleton private constructor(private val data: String) {

    companion object {
        @Volatile
        private var instance: ThreadSafeSingleton? = null

        fun getInstance(data: String): ThreadSafeSingleton {
            var result = instance

            if (result == null) {
                synchronized(this) {
                    result = instance
                    if (result == null) {
                        result = ThreadSafeSingleton(data)
                        instance = result
                    }
                }
            }
            return result!!
        }
    }

    fun getData(): String = data
}

// kotlin simple alternative with lazy delegate
object LazySingleton {
    val data: String by lazy { "Singleton Data" }
}

fun main() {
    testThreadSafeSingleton()
    testLazySingleton()
}

fun testThreadSafeSingleton() {
    // Create two threads to test thread-safe initialization
    val thread1 = Thread {
        val singleton1 = ThreadSafeSingleton.getInstance("Thread 1 Data")
        println("Thread 1 Singleton Data: ${singleton1.getData()}")
    }

    val thread2 = Thread {
        val singleton2 = ThreadSafeSingleton.getInstance("Thread 2 Data")
        println("Thread 2 Singleton Data: ${singleton2.getData()}")
    }

    // Start the threads
    thread1.start()
    thread2.start()

    // Wait for threads to complete
    thread1.join()
    thread2.join()

    // Verify that the singleton instance is consistent across all threads
    val mainThreadSingleton = ThreadSafeSingleton.getInstance("Main Data")
    println("Main Thread Singleton Data: ${mainThreadSingleton.getData()}")
}

fun testLazySingleton() {
    // Access the lazy singleton
    println("LazySingleton Data: ${LazySingleton.data}")
}
